plugins/module_utils/mh/mixins/state.py

```diff
@@ -11,29 +11,21 @@ class StateMixin(object):
     state_param = 'state'
     default_state = None
 
-    def _state(self):
-        state = self.module.params.get(self.state_param)
+    @property
+    def _state_value(self):
+        state = self.vars.get(self.state_param)
         return self.default_state if state is None else state
 
     def _method(self, state):
         return "{0}_{1}".format(self.state_param, state)
 
     def __run__(self):
-        state = self._state()
-        self.vars.state = state
-
-        # resolve aliases
-        if state not in self.module.params:
-            aliased = [name for name, param in self.module.argument_spec.items() if state in param.get('aliases', [])]
-            if aliased:
-                state = aliased[0]
-                self.vars.effective_state = state
```

Comment:
Is this the code you say is extraneous? If it is: why?

Reply:
Because I made a lame mistake: in line 22, we load the *value* of the state passed by the module user into the variable `state`, then in line 26 we search for a *module parameter* named after that state value, which makes no sense - no parameter `absent` or `present`, those are values of the `state` param.

And the logic for aliases is also unnecessary - is `state` has an alias, the module parsing will always set `module.params['state']`.

Reply:
I agree that this code doesn't really make sense, but it has potential side-effects, and so removing it can potentially break something.

Also about the alias logic: if say the state is `foo`, and you have an option that's called `bar` with alias `foo`, and the user did specify `bar` but not `foo` as a module parameter, this will change `state` to `bar`. So removing this code will suddenly result in another value of `state`.

I think this piece of code should be deprecated and then removed in a later major release. I don't think / hope anyone actually relies on it, but if someone does, just removing it is a bad idea.

Reply:
If state is `foo`, the test in line 26 will fail and it will never enter this block of code, because there is no parameter named `foo`.

On top of that, if the option is `bar` and the alias is `foo`, a task like:
```yaml
- mymodule:
    foo: 3
```
when parsed by Ansible, will NOT generate an element `module.params['foo']`, it will resolve the alias and create `module.params.['bar'] = 3`. So, from the module point of view, we do not need to process the parameter `foo`, only `bar`. 

And finally, the variable `state` is used at some point as the _value_ of the `state` parameter, and a couple of lines down it is used as the _name_ of the parameter. I really don't know where my head was when I wrote that crap. I will bang my head on the wall for a couple of weeks, be back in a moment. ;-)

Reply:
> If state is foo, the test in line 26 will fail and it will never enter this block of code, because there is no parameter named foo.

That's not true. If `state == foo` and there is no module option named `foo`, as in my example, the code block will get executed.

> when parsed by Ansible, will NOT generate an element module.params['foo'], it will resolve the alias and create module.params.['bar'] = 3. So, from the module point of view, we do not need to process the parameter foo, only bar.

That's not true, at least for Ansible 2.9 and ansible-base 2.10; see https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/module_utils/common/parameters.py#L156-L187 for the code which handles aliases. The code was changed for ansible-core 2.11, so potentially it behaves differently from ansible-core 2.11 on.

> And finally, the variable state is used at some point as the value of the state parameter, and a couple of lines down it is used as the name of the parameter. I really don't know where my head was when I wrote that crap. I will bang my head on the wall for a couple of weeks, be back in a moment. ;-)

:rofl:

Reply:
@russoz how shall we continue here?

Reply:
Still inflicting self-punishment for that :rofl: . Kidding, I am going back to this over the weekend.

Reply:
Changed the approach, I believe it is clearer now.